Always render suggestions – fix #133

ThibWeb opened the pull request:
Hey hey,

this is a PR to fix #133. It implements a new `alwaysRenderSuggestions` boolean prop, as we discussed. When `alwaysRenderSuggestions={true}`, the checks to decide whether to render the suggestions are bypassed. Also comes with some documentation on this new prop, and relevant tests.

I'd be keen to get a review of whether [the tests are relevant and thorough enough](https://github.com/ThibWeb/react-autosuggest/blob/feature/always-render-suggestions/test/plain-list/AutosuggestApp.test.js#L431), and make sure this new prop [is used in the right places](https://github.com/moroshko/react-autosuggest/pull/155/commits/0b0a86cebe3391adabf995bb02323ed478381877).

Also two things occurred to me while writing this PR:

- Having an example for this would only be relevant if it came with the kind of UI this is useful for. Do we want to make one? (another way to see this would be to have a list of "projects `react-autosuggest` is used on" somewhere).
- It might or might not be be obvious that in order for this to be useful, you need to have a list of suggestions to display _even when the user hasn't typed anything yet_ (otherwise there's nothing to display, doh). Does this need to be stated in the docs?

moroshko commented on test/plain-list/AutosuggestApp.js:
```diff
@@ -48,6 +48,13 @@ export const shouldRenderSuggestions = sinon.spy(value => {
+export const toggleAlwaysRenderSuggestions = sinon.spy((toggle, callback) => {
+  app.setState({
+    alwaysRenderSuggestions: toggle,
+    suggestions: getMatchingLanguages(app.state.value || 'elm')
```
👍 for "elm", but why `|| 'elm'` is required?

ThibWeb replied:
For those tests to be useful it is necessary to get suggestions even when there's no value in the input – that's why I use `elm` here. It's not what would happen in a real-world app but does the same job for the tests' purpose.

I added a comment to make this clearer.

moroshko commented on src/Autosuggest.js:
```diff
@@ -176,13 +179,14 @@ class Autosuggest extends Component {
       suggestions, renderSuggestion, inputProps, shouldRenderSuggestions,
       onSuggestionSelected, multiSection, renderSectionTitle, id,
-      getSectionSuggestions, focusInputOnSuggestionClick, theme, isFocused,
+      getSectionSuggestions, focusInputOnSuggestionClick, theme,isFocused,
```
really?

ThibWeb replied:
Erm. Feels like ESLint should have a rule for that but I haven't been able to find it.

moroshko commented on README.md:
```diff
@@ -328,6 +329,13 @@ function shouldRenderSuggestions(value) {
+<a name="alwaysRenderSuggestionsProp"></a>
+#### alwaysRenderSuggestions (optional)
+
+By default, suggestions are rendered only if the input field is under focus.
```
"under focus" => "focused" (to keep the terminology consistent)

moroshko commented on README.md:
```diff
@@ -328,6 +329,13 @@ function shouldRenderSuggestions(value) {
+
+By default, suggestions are rendered only if the input field is under focus.
+
+If you'd like to _always_ display suggestions, set `alwaysRenderSuggestions={true}`.
```
Shall we also add a note in the `shouldRenderSuggestions` section along the lines of:
> When `shouldRenderSuggestions` returns `true`, suggestions will be rendered only when the input field is focused. If you would like to render suggestions regardless of whether the input field is focused or not, set `alwaysRenderSuggestions={true}` (`shouldRenderSuggestions` is ignored in this case).

?

ThibWeb replied:
I was wondering if having the two documentation sections next to one another would be enough. Doesn't hurt to make this info more explicit.

moroshko commented on test/plain-list/AutosuggestApp.js:
```diff
@@ -48,6 +48,13 @@ export const shouldRenderSuggestions = sinon.spy(value => {
 
+export const toggleAlwaysRenderSuggestions = sinon.spy((toggle, callback) => {
+  app.setState({
+    alwaysRenderSuggestions: toggle,
```
"toggle" confuses me here, as I normally think about toggle as "reverse the current state". How about `setAlwaysRenderSuggestions` instead?

ThibWeb replied:
Sounds good to me!

moroshko commented:
Overall looks great! Just a few minor comments above.

ThibWeb commented:
As those tweaks were quite minor I just rebased them into my previous commits:

- Fix spacing in variable declaration list
- `toggleAlwaysRenderSuggestions` -> `setAlwaysRenderSuggestions`
- Documentation changes
- Comment to make the test helper's intent clearer.

moroshko commented on src/Autosuggest.js:
```diff
@@ -78,11 +79,13 @@ class Autosuggest extends Component {
       const { value } = inputProps;
 
-      if (isCollapsed && lastAction !== 'click' && lastAction !== 'enter' &&
+      if (alwaysRenderSuggestions || isCollapsed && lastAction !== 'click' && lastAction !== 'enter' &&
```
Do we really need `alwaysRenderSuggestions ||` here? 
All tests pass without it, so I wonder if we can just remove it, or are we missing a test here?

moroshko commented:
@ThibWeb Thanks for the tweaks!

I think I found an issue with the implementation.

I just added `alwaysRenderSuggestions={true}` to the CustomRenderer example (also had to change `[]` to `people` [here](https://github.com/moroshko/react-autosuggest/blob/master/demo/src/components/App/components/Examples/components/CustomRender/CustomRender.js#L18)). 
Suggestions are initially rendered, as expected. 
However, when I hit `Down`, the first suggestion isn't getting focus. Changing [this line](https://github.com/moroshko/react-autosuggest/blob/master/src/Autosuggest.js#L219) to:
```js
if (!alwaysRenderSuggestions && isCollapsed) ...
```
solves the issue, but it doesn't feel right. `isCollapsed` is `true` even though suggestions are shown.

Basically, we have a prop and a state that control the view:
* `alwaysRenderSuggestions` - prop
* `isCollapsed` - state

I feel that the prop should set the initial state. Then, `alwaysRenderSuggestions` should be used to make sure that `isCollapsed` stays `true` no matter what. This way, `isCollapsed` will be `true` if and only if suggestions are shown.

What do you think?

ThibWeb commented:
Yes, I've had some trouble understanding how best to interact with `isCollapsed` – doesn't surprise me that I got it wrong. I'll revisit this and probably create a standalone example, as much as a help when developing as to demonstrate what the prop is for.

bradbirnbaum commented:
Checking in on this pull - do we expect to take it soon?

moroshko commented:
@ThibWeb Have you had a chance to revisit this PR?

bradbirnbaum commented:
Checking in to see if we expect this to be released soon.   Thanks

ThibWeb commented:
Hey there – sorry for the delay, I was on holidays for a few weeks.

I still intend to work on this, and will commit to get back with a better PR _this weekend_. Hopefully this will be releasable.

jxodwyer commented:
@ThibWeb just wondering the status of this PR? I'd be happy to jump in and help out if you're running into issues or strapped for time.

ThibWeb commented:
Hey @jxodwyer, strapped for time indeed. If you want to pick this up by all means please go for it! Two weeks ago I got back onto this and made the UI example. That makes it easier to see what is going, and where it fails to work.

Useful links:

- Compiled UI example, for reference: https://rawgit.com/ThibWeb/react-autosuggest/demo/arr/demo/dist/index.html
- Branch I'm working from https://github.com/ThibWeb/react-autosuggest/tree/feature/always-render-suggestions

Next things to do IMHO:

- [ ] Look at [moroshko's suggestions/hints](https://github.com/moroshko/react-autosuggest/pull/155#issuecomment-216216991] above.
- [ ] Debug the current implementation with those hints and the examples as visual aids
- [ ] Make sure existing test suite still passes
- [ ] Add more tests for the cases that do not work here
- [ ] Update PR 🚀 
- [ ] Figure out if this example needs a Codepen link (under moroshko's account like the others?)

I don't expect this all to take a crazy amount of time (1-2 hours if you've seen the `react-autosuggest` code before).

I've added you as a contributor to https://github.com/ThibWeb/react-autosuggest, it's up to you if you prefer to use this as is or start from it in another repo 👌.

shprink commented:
This feature is much needed indeed! Thanks for your hard work guys, can't wait to see that merged!

ThibWeb commented:
@moroshko I have updated the PR, could you have a look?

I followed your [suggestions](https://github.com/moroshko/react-autosuggest/pull/155#issuecomment-216216991), `alwaysRenderSuggestions` is now used to set the default value of `isCollapsed`, and to set `shouldRenderSuggestions` to always return true. Ideally I would have wanted to completely remove any references to `alwaysRenderSuggestions` in the `Autosuggest` component, but there are a few instances where `isCollapsed` and `shouldRenderSuggestions` are not the only criteria to determine what is rendered:

- https://github.com/ThibWeb/react-autosuggest/blob/master/src/Autosuggest.js#L85
- https://github.com/ThibWeb/react-autosuggest/blob/master/src/Autosuggest.js#L185

Those extra checks are useful, but having them within the `Autosuggest` component makes the code harder to change – there are too many places in the code at different levels of abstraction that are in charge of determining whether suggestions are shown or not. Ideally I think that a single `isOpen`/`isCollapsed` prop should be the one and only source of truth for whether suggestions are shown or not. A way to get towards this would be to change how `isCollapsed` is set [`mapDispatchToProps`](https://github.com/ThibWeb/react-autosuggest/blob/master/src/Autosuggest.js#L10), but that's outside of the scope of this PR.

Other than that, I added the relevant tests to check for:

- the unwanted behavior that the previous version of this PR produced.
- the wanted behavior when selecting a suggestion

And I updated https://rawgit.com/ThibWeb/react-autosuggest/demo/arr/demo/dist/index.html to make reviewing this easier.

The last thing that's missing would be a CodePen link – not too sure how to proceed here, since this isn't something we can collaborate on via PRs?

rickhanlonii commented:
@moroshko just a heads up that merging this will cause some conflicts with https://github.com/moroshko/react-autosuggest/pull/165. After this is merged I'll update #165 and add any tests necessary for both props to be active at once.

ThibWeb commented:
Just a heads up that if someone wants to use this sooner rather than later, I have a branch where this work is compiled and npm installable: https://github.com/ThibWeb/react-autosuggest/tree/feature/arr-compiled

I'd advise _against_ relying on someone else's branch in a fork for something serious, but you can easily fork it to your own GitHub account to be more certain that it will stay how it is, and then `npm install --save react-autossugest@git://github.com/ThibWeb/react-autosuggest.git#feature/arr-compiled` with your own user / repo / branch.

moroshko commented:
@ThibWeb I'm keen to review and merge this PR. Do you mind resolving the merge conflicts? (sorry, they appeared after I merged #165)

carloscuatin commented:
@ThibWeb resolve conflicts please 😀

shprink commented:
🙇  @ThibWeb I will pay you a beer when you come to France if you resolve the conflicts :)

ThibWeb commented:
Wuhu, conflicts merged! They were quite straightforward, only had to add the `alwaysRenderSuggestions` vars where they belong.

Regarding `focusFirstSuggestion` – when both this and `alwaysRenderSuggestions` are turned on, at the moment it only puts the autosuggest focus on the first suggestion _when the user moves focus to the autosuggest field_. Focusing one of the suggestions even though the autosuggest does not have focus feels strange to me (happy to be told I'm wrong).

The running example at https://rawgit.com/ThibWeb/react-autosuggest/demo/arr/demo/dist/index.html is up to date. @moroshko this might not pass the linting since your latest config updates – I tried to make it work but `eslint-plugin-react@6.0.0-alpha.1` keeps complaining about unfound rule definitions.

LuckUY commented:
Hi @moroshko,

Any update on this new feature? When do you think it will be merged?

Thanks

carloscuatin commented:
nice @moroshko 🎉

shprink commented:
Sending <3 to all of you!

LuckUY commented:
@moroshko 

I cannot see the new "alwaysRenderSuggestions" prop on Autosuggest.js (v 3.9.0)
Did you publish a new version?

moroshko commented:
@LuckUY Not yet. I hope to publish new version sometime this week.

moroshko commented:
`alwaysRenderSuggestions` is available in [v4.0.0](https://github.com/moroshko/react-autosuggest/releases/tag/v4.0.0) now.
